src/dmd/root/timeit.d

```diff
@@ -0,0 +1,128 @@
+/**
+ * Development utility for tracking the amount of time a section of code takes.
+ *
+ * Copyright:   Copyright (C) 1999-2021 by The D Language Foundation, All Rights Reserved
+ * Authors:     Stefan Koch
+ * License:     $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
+ * Source:      $(LINK2 https://github.com/dlang/dmd/blob/master/src/dmd/root/timeit.d, _timeit.d)
+ * Documentation:  https://dlang.org/phobos/dmd_root_timeit.html
+ * Coverage:    https://codecov.io/gh/dlang/dmd/src/master/src/dmd/root/timeit.d
+ */
+
+module dmd.root.timeit;
+
+/// Accumulates the ticks for this section of code into the provided counter
+struct TimeIt
+{
+    ulong *ticks_var;
+    ulong begin;
+
+    this(ulong* var)
+    {
+        pragma(inline, true);
+        ticks_var = var;
+        begin = tsc();
+    }
+
+    @disable this(this);
+    @disable ref @system TimeIt opAssign(TimeIt p) return;
+
+    ~this()
+    {
+        pragma(inline, true);
+        *ticks_var += (tsc() - begin);
+    }
+}
+
+version (D_InlineAsm_X86_64)
+{
+    static ulong tsc()
+    {
+        asm nothrow @nogc
+        {
+            naked;
+            mfence;      // force serialisation (this also makes it kinda slow :-/)
```

Comment:
Do you want to serialize memory or the instruction stream? just in case it's the latter:
The (Instruction) serializing behaviour of the X86 fence instructions, is - for at least some CPUs - dependent on the vendor: If I recall correctly Intel recommend `lfence` whereas `mfence` works on AMD.

Also of note: `rdtscp` has been around for a very long time, so you could use that if needed (Although LDC extended asm doesn't support it, for reference)

Reply:
However, I am fairly sceptical either the fence or my comments will have any relevance to the utility of this code in practice.

Reply:
Before any more discussion about this, note that druntime contains exactly this piece of code, although not limited to x64 *and* not exposed to the public - https://github.com/dlang/druntime/blob/b42243287745b506004e0d48a7efacd4bc5481ae/src/rt/trace.d#L865-L932.

Reply:
@kinke thoughts on exposing it to the public? Seems useful enough. 

P.S. I see that the kernel source is mentioned as a reference, it may be of note for anyone reading this that Intel have some good documentation (both in their big manuals and stand-alone) on how to use the time stamp counters effectively.

Reply:
I think exposing it would be useful, yes, just like more stuff in `rt.*` (e.g., reading sections from binaries etc.). But the 'problem' is that we can only use it in the compiler here when bumping the min host compiler version accordingly, which takes years. But for such a little dev utility module, it could work (`static if (__VERSION__ >= 2098)` or `__traits(compiles)` etc.).

Reply:
The second best time to start the bootstrap process is now though, right? 

Also does it actually (have to) take years? I understand the importance of the process but presumably it's not impossible that we get outcompeted by other languages because we hold ourselves back with good but flawed intentions. I don't have to actually do releases, however, so maybe I'm being naive. 

On the subject of development speed, would a druntime+dmd monorepo (of sorts) speed this process up in the future? @ibuclaw @Geod24 ?

Reply:
> The second best time to start the bootstrap process is now though, right?

Yeah, definitely. ;)

Regarding min host compiler version: I think that's mainly a non-DMD thing - being able to use old host compilers does have advantages, especially when bootstrapping D on new (read: non-x86) platforms. GDC is still at v2.076 (at least from the druntime side), so being able to bootstrap LDC via a GDC already ported to and working on a new platform is very convenient. [LDC can still be bootstrapped with LDC v0.17 (D 2.068!) - at least on 64-bit platforms; but that won't work for much longer.]

Reply:
@kinke the difference to the version in druntime is that I don't hardcode 4byte movs because it's written for amd64.
I agree that we should use druntime in the longterm but seeing that it's private and not particularly optimized asm.
This is really about going the first step into the direction of having profiling easily available in the dmd codebase.